p2p/protocols/accounting.go

```diff
@@ -122,17 +119,18 @@ func SetupAccountingMetrics(reportInterval time.Duration, path string) *Accounti
 }
 
 // Send takes a peer, a size and a msg and
-//   - calculates the cost for the local node sending a msg of size to peer using the Prices interface
+//   - calculates the cost for the local node sending a msg of size to peer querying the message for its price
 //   - credits/debits local node using balance interface
 func (ah *Accounting) Send(peer *Peer, size uint32, msg interface{}) error {
```

Comment:
As this function can only succeed by passing a PricedMessage, we can make it a compile-time error if no priced-message is passed instead of a run-time error when we replace `msg interface{}` with `msg PricedMessage`.

I just tried this locally and got an error

```
network/retrieval/retrieve.go:115:15: cannot use protocols.NewAccounting(balance) (type *protocols.Accounting) as type protocols.Hook in assignment:
	*protocols.Accounting does not implement protocols.Hook (wrong type for Send method)
		have Send(*protocols.Peer, uint32, protocols.PricedMessage) error
		want Send(*protocols.Peer, uint32, interface {}) error
```
which is resolvable by updating the Hook interface (`p2p/protocols/protocol.go`).

Additionally, this will make the body of the function also more concise.

Same holds for the Receive implementation in `p2p/protocols.go/accounting.go:139` => change the function to accept only a PricedMessage type.

Reply:
keep in mind that both of the `Send` and `Receive` functions return `error` variables. 

when you pass a message which does not implement the `PricedMessage` interface, the code returns `nil`. this means **success**—there is no price, therefore no accounting to be done, therefore no error at this level. 

in other words, we should go ahead and process the protocol message if it's not a priced one. 

what's more: by only accepting priced messages, we would be blocking every other message from that protocol from being processed. if the protocol has an accounting hook set, it will call `Send` and `Receive` for _all_ its messages, and all should be able to be processed whether they are priced or not.

i would like to know more about this error you are getting, can you give more information?

Reply:
@Eknir, if you do this, you need to implement Price for every message of the protocol. 
Plus `Send` and `Receive` are generic protocol level hooks with generic type signature, accounting is just one such hook (since no other exists, this argument is somewhat weak, but nontheless in line with the design.)

Reply:
Thanks @mortelli and @victor . I was somewhat mislead because I had the idea that both of these functions were not intended to be called by a message which does not need accounting. If that is not the case, my argument obviously does not hold.

I dived a bit more in the code and I see that indeed, the accounting.Send() function is called with many different message types () `p2p/protocol/protocol.go` and that it because of this is needed that the function can be called also by messages for which no price is defined.

@mortelli , the error I am getting is because I changed the type of `msg` in the Send function, but I did not change it in the in the protocols.Hook (`p2p/protocols/protocols.go:117`) definition.
This makes `Accounting` (`p2p/protocols/accounting.go`) to not anymore implement the Hook interface, leading to the error.